File: src/main/services/mod-manager/shader-fixes.test.ts
import assert from "node:assert/strict";
import { createHash } from "node:crypto";
import os from "node:os";
import path from "node:path";

import fse from "fs-extra";
import { afterEach, beforeEach, describe, it, vi } from "vitest";

import type { NahidaDesktop } from "../..";
import type { ModLibraryService } from "./library";
import type { ShaderFixesProcessedFile } from "./shader-fixes";

import { ModShaderFixesService, SHADER_FIXES_MOD_MARKER_FILE } from "./shader-fixes";

const fastGlobCalls = vi.hoisted(
    () => [] as { pattern: string | string[]; cwd: string | undefined }[],
);
const atomicWriteState = vi.hoisted(() => ({ calls: 0, failOnCall: null as number | null }));

vi.mock("fast-glob", async (importOriginal) => {
    const original = await importOriginal<typeof import("fast-glob")>();
    const fastGlob = Reflect.get(original, "default") as typeof import("fast-glob");
    return {
        ...original,
        default: new Proxy(fastGlob, {
            apply(target, thisArg, args) {
                fastGlobCalls.push({
                    pattern: args[0] as string | string[],
                    cwd: (args[1] as { cwd?: string } | undefined)?.cwd,
                });
                return Reflect.apply(target, thisArg, args);
            },
        }),
    };
});

vi.mock("write-file-atomic", async (importOriginal) => {
    const original = await importOriginal<typeof import("write-file-atomic")>();
    const writeFileAtomic = Reflect.get(original, "default") as typeof import("write-file-atomic");
    return {
        ...original,
        default: new Proxy(writeFileAtomic, {
            apply(target, thisArg, args) {
                atomicWriteState.calls += 1;
                if (atomicWriteState.calls === atomicWriteState.failOnCall) {
                    throw new Error("OWNER_INDEX_WRITE_FAILED");
                }
                return Reflect.apply(target, thisArg, args);
            },
        }),
    };
});

describe("ModShaderFixesService", () => {
    let rootPath: string;
    let importerPath: string;
    let modsPath: string;
    let service: ModShaderFixesService;
    let importers: { key: string; importerFolder: string }[];
    let games: { game: string; modFolderPath: string; importer: string }[];

    beforeEach(async () => {
        rootPath = await fse.mkdtemp(path.join(os.tmpdir(), "nahida-shader-fixes-"));
        importerPath = path.join(rootPath, "Importer");
        modsPath = path.join(importerPath, "Mods");
        await fse.ensureDir(modsPath);
        importers = [{ key: "Importer", importerFolder: importerPath }];
        games = [{ game: "Test", modFolderPath: modsPath, importer: "Importer" }];

        service = new ModShaderFixesService(
            {
                service: {
                    xxmi: {
                        getEnabledImporters: () => importers,
                    },
                },
                lib: {
                    utils: {
                        getFileHash: async (filePath: string) =>
                            createHash("sha256")
                                .update(await fse.readFile(filePath))
                                .digest("hex"),
                    },
                },
                logger: { error: vi.fn() },
            } as unknown as NahidaDesktop,
            {
                games: async () => games,
            } as unknown as ModLibraryService,
        );

        fastGlobCalls.length = 0;
        atomicWriteState.calls = 0;
        atomicWriteState.failOnCall = null;
    });

    afterEach(async () => {
        await fse.remove(rootPath);
    });

    it("uses the central owner index without scanning when disabling", async () => {
        const modPath = path.join(modsPath, "Group", "Mod");
        await Promise.all(
            ["a.ini", "nested/b.ini", "nested/c.ini"].map(async (file) => {
                await fse.outputFile(path.join(modPath, "ShaderFixes", file), file);
            }),
        );
        await service.handleShaders(modPath, true);
        assert.equal(atomicWriteState.calls, 2);
        await fse.remove(path.join(modPath, "ShaderFixes"));
        fastGlobCalls.length = 0;

        await service.handleShaders(modPath, false);

        assert.deepEqual(fastGlobCalls, []);
        assert.equal(await fse.pathExists(path.join(importerPath, "ShaderFixes", "a.ini")), false);
        assert.equal(await fse.pathExists(path.join(modPath, SHADER_FIXES_MOD_MARKER_FILE)), false);
        assert.deepEqual(
            await fse.readJson(
                path.join(importerPath, "ShaderFixes", SHADER_FIXES_MOD_MARKER_FILE),
            ),
            { version: 1, targets: {} },
        );
    });

    it("preserves a shader fix while another mod owns it", async () => {
        const firstModPath = path.join(modsPath, "Group", "First");
        const secondModPath = path.join(modsPath, "Group", "Second");
        await Promise.all(
            [firstModPath, secondModPath].map(async (modPath) => {
                await fse.outputFile(path.join(modPath, "ShaderFixes", "shared.ini"), "shared");
                await service.handleShaders(modPath, true);
            }),
        );

        await service.handleShaders(firstModPath, false);
        assert.equal(
            await fse.pathExists(path.join(importerPath, "ShaderFixes", "shared.ini")),
            true,
        );

        await service.handleShaders(secondModPath, false);
        assert.equal(
            await fse.pathExists(path.join(importerPath, "ShaderFixes", "shared.ini")),
            false,
        );
    });

    it("preserves a shader fix that changed after it was copied", async () => {
        const modPath = path.join(modsPath, "Group", "Mod");
        const targetPath = path.join(importerPath, "ShaderFixes", "changed.ini");
        await fse.outputFile(path.join(modPath, "ShaderFixes", "changed.ini"), "original");
        await service.handleShaders(modPath, true);
        await fse.writeFile(targetPath, "changed");

        await service.handleShaders(modPath, false);

        assert.equal(await fse.readFile(targetPath, "utf8"), "changed");
        assert.equal(await fse.pathExists(path.join(modPath, SHADER_FIXES_MOD_MARKER_FILE)), false);
    });

    it("migrates existing manifests by scanning only the current importer mods", async () => {
        const firstModPath = path.join(modsPath, "Group", "First");
        const secondModPath = path.join(modsPath, "Group", "Second");
        const targetPath = path.join(importerPath, "ShaderFixes", "shared.ini");
        const hash = createHash("sha256").update("shared").digest("hex");
        await fse.outputFile(targetPath, "shared");
        await Promise.all(
            [
                { modPath: firstModPath, modKey: "first" },
                { modPath: secondModPath, modKey: "second" },
            ].map(async ({ modPath, modKey }) => {
                await fse.outputJson(path.join(modPath, SHADER_FIXES_MOD_MARKER_FILE), {
                    version: 1,
                    modKey,
                    files: [
                        {
                            file: "shared.ini",
                            targetPath,
                            targetKey: "legacy-target-key",
                            hash,
                        },
                    ],
                });
            }),
        );
        const otherImporterPath = path.join(rootPath, "OtherImporter");
        const otherModsPath = path.join(otherImporterPath, "Mods");
        await fse.ensureDir(otherModsPath);
        importers.push({ key: "OtherImporter", importerFolder: otherImporterPath });
        games.push({ game: "Other", modFolderPath: otherModsPath, importer: "OtherImporter" });
        fastGlobCalls.length = 0;

        await service.handleShaders(firstModPath, false);

        assert.deepEqual(fastGlobCalls, [
            { pattern: `**/${SHADER_FIXES_MOD_MARKER_FILE}`, cwd: modsPath },
        ]);
        assert.equal(await fse.pathExists(targetPath), true);
        assert.deepEqual(
            await fse.readJson(
                path.join(importerPath, "ShaderFixes", SHADER_FIXES_MOD_MARKER_FILE),
            ),
            {
                version: 1,
                targets: { "shared.ini": { hash, owners: ["second"] } },
            },
        );

        fastGlobCalls.length = 0;
        await service.handleShaders(secondModPath, false);
        assert.deepEqual(fastGlobCalls, []);
        assert.equal(await fse.pathExists(targetPath), false);
    });

    it("cleans the original importer after an enabled mod is moved", async () => {
        const originalModPath = path.join(modsPath, "Group", "Mod");
        const originalTargetPath = path.join(importerPath, "ShaderFixes", "moved.ini");
        await fse.outputFile(path.join(originalModPath, "ShaderFixes", "moved.ini"), "moved");
        await service.handleShaders(originalModPath, true);

        const otherImporterPath = path.join(rootPath, "OtherImporter");
        const otherModsPath = path.join(otherImporterPath, "Mods");
        const movedModPath = path.join(otherModsPath, "Group", "Mod");
        await fse.ensureDir(path.dirname(movedModPath));
        importers.push({ key: "OtherImporter", importerFolder: otherImporterPath });
        games.push({ game: "Other", modFolderPath: otherModsPath, importer: "OtherImporter" });
        await fse.move(originalModPath, movedModPath);
        await fse.remove(path.join(importerPath, "ShaderFixes", SHADER_FIXES_MOD_MARKER_FILE));
        fastGlobCalls.length = 0;

        await service.handleShaders(movedModPath, false);

        assert.deepEqual(fastGlobCalls, [
            { pattern: `**/${SHADER_FIXES_MOD_MARKER_FILE}`, cwd: modsPath },
        ]);
        assert.equal(await fse.pathExists(originalTargetPath), false);
        assert.equal(
            await fse.pathExists(path.join(movedModPath, SHADER_FIXES_MOD_MARKER_FILE)),
            false,
        );
        assert.deepEqual(
            await fse.readJson(
                path.join(importerPath, "ShaderFixes", SHADER_FIXES_MOD_MARKER_FILE),
            ),
            { version: 1, targets: {} },
        );
        assert.equal(
            await fse.pathExists(
                path.join(otherImporterPath, "ShaderFixes", SHADER_FIXES_MOD_MARKER_FILE),
            ),
            false,
        );
    });

    it("rebuilds a corrupted owner index before deleting files", async () => {
        const modPath = path.join(modsPath, "Group", "Mod");
        const targetPath = path.join(importerPath, "ShaderFixes", "legacy.ini");
        const hash = createHash("sha256").update("legacy").digest("hex");
        await fse.outputFile(targetPath, "legacy");
        await fse.outputJson(path.join(modPath, SHADER_FIXES_MOD_MARKER_FILE), {
            version: 1,
            modKey: "legacy",
            files: [
                {
                    file: "legacy.ini",
                    targetPath,
                    targetKey: "legacy-target-key",
                    hash,
                },
            ],
        });
        await fse.writeFile(
            path.join(importerPath, "ShaderFixes", SHADER_FIXES_MOD_MARKER_FILE),
            "invalid",
        );
        fastGlobCalls.length = 0;

        await service.handleShaders(modPath, false);

        assert.deepEqual(fastGlobCalls, [
            { pattern: `**/${SHADER_FIXES_MOD_MARKER_FILE}`, cwd: modsPath },
        ]);
        assert.equal(await fse.pathExists(targetPath), false);
    });

    it("removes owner state when an enable operation is rolled back", async () => {
        const modPath = path.join(modsPath, "Group", "Mod");
        const targetPath = path.join(importerPath, "ShaderFixes", "rollback.ini");
        await fse.outputFile(path.join(modPath, "ShaderFixes", "rollback.ini"), "rollback");

        const processedFiles = await service.handleShaders(modPath, true);
        await service.rollbackEnabledShaders(modPath, processedFiles);

        assert.equal(await fse.pathExists(targetPath), false);
        assert.deepEqual(
            await fse.readJson(
                path.join(importerPath, "ShaderFixes", SHADER_FIXES_MOD_MARKER_FILE),
            ),
            { version: 1, targets: {} },
        );
        assert.equal(await fse.pathExists(path.join(modPath, SHADER_FIXES_MOD_MARKER_FILE)), false);
    });

    it("can roll back when saving owner state fails during enable", async () => {
        const modPath = path.join(modsPath, "Group", "Mod");
        const targetPath = path.join(importerPath, "ShaderFixes", "rollback.ini");
        await fse.outputFile(path.join(modPath, "ShaderFixes", "rollback.ini"), "rollback");
        atomicWriteState.failOnCall = 2;

        const error = await service
            .handleShaders(modPath, true)
            .catch((failure: unknown) => failure);
        assert.equal(error instanceof Error, true);
        const processedFiles = (error as { processedFiles: ShaderFixesProcessedFile[] })
            .processedFiles;
        assert.equal(processedFiles.length, 1);

        atomicWriteState.failOnCall = null;
        await service.rollbackEnabledShaders(modPath, processedFiles);

        assert.equal(await fse.pathExists(targetPath), false);
        assert.deepEqual(
            await fse.readJson(
                path.join(importerPath, "ShaderFixes", SHADER_FIXES_MOD_MARKER_FILE),
            ),
            { version: 1, targets: {} },
        );
    });

    it("does not copy the reserved owner index from a mod ShaderFixes folder", async () => {
        const modPath = path.join(modsPath, "Group", "Mod");
        await fse.outputFile(path.join(modPath, "ShaderFixes", "actual.ini"), "actual");
        await fse.outputFile(
            path.join(modPath, "ShaderFixes", SHADER_FIXES_MOD_MARKER_FILE),
            "reserved",
        );

        await service.handleShaders(modPath, true);

        assert.deepEqual(
            await fse.readJson(
                path.join(importerPath, "ShaderFixes", SHADER_FIXES_MOD_MARKER_FILE),
            ),
            {
                version: 1,
                targets: {
                    "actual.ini": {
                        hash: createHash("sha256").update("actual").digest("hex"),
                        owners: [
                            (await fse.readJson(path.join(modPath, SHADER_FIXES_MOD_MARKER_FILE)))
                                .modKey,
                        ],
                    },
                },
            },
        );
    });

    it("does not scan the mod when no manifest exists", async () => {
        const modPath = path.join(modsPath, "Group", "Mod");
        await fse.outputFile(path.join(modPath, "ShaderFixes", "unused.ini"), "unused");
        fastGlobCalls.length = 0;

        await service.handleShaders(modPath, false);

        assert.deepEqual(fastGlobCalls, []);
    });
});

File: internal/mod/shader_fixes_test.go
package mod

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

func TestShaderFixesUsesOwnerIndexWithoutScanningWhenDisabling(t *testing.T) {
	t.Parallel()
	h := newShaderFixesHarness(t)
	modPath := filepath.Join(h.modsPath, "Group", "Mod")
	for _, file := range []string{"a.ini", filepath.Join("nested", "b.ini"), filepath.Join("nested", "c.ini")} {
		writeFile(t, filepath.Join(modPath, "ShaderFixes", file), filepath.Base(file))
	}
	if _, err := h.service.HandleShaders(modPath, true); err != nil {
		t.Fatal(err)
	}
	if h.service.ownerIndexWrites != 2 {
		t.Fatalf("owner index writes = %d, want 2", h.service.ownerIndexWrites)
	}
	if err := os.RemoveAll(filepath.Join(modPath, "ShaderFixes")); err != nil {
		t.Fatal(err)
	}
	h.service.globCalls = nil

	if _, err := h.service.HandleShaders(modPath, false); err != nil {
		t.Fatal(err)
	}
	if len(h.service.globCalls) != 0 {
		t.Fatalf("glob calls = %#v, want none", h.service.globCalls)
	}
	if exists(filepath.Join(h.importerPath, "ShaderFixes", "a.ini")) {
		t.Fatal("copied shader should be removed")
	}
	if exists(filepath.Join(modPath, shaderFixesModMarkerFile)) {
		t.Fatal("mod marker should be removed")
	}
	assertOwnerIndex(t, filepath.Join(h.importerPath, "ShaderFixes", shaderFixesModMarkerFile), map[string]any{
		"version": float64(1), "targets": map[string]any{},
	})
}

func TestNormalizeShaderFixesOwnerTargetKeyOnlyRejectsParentSegments(t *testing.T) {
	t.Parallel()

	service := &ShaderFixes{}
	if got := service.normalizeShaderFixesOwnerTargetKey("nested/x..y.ini"); got == nil || *got != "nested/x..y.ini" {
		t.Fatalf("ordinary double-dot filename = %#v", got)
	}
	for _, path := range []string{"../escape.ini", "nested/../escape.ini", `nested\..\escape.ini`} {
		if got := service.normalizeShaderFixesOwnerTargetKey(path); got != nil {
			t.Fatalf("parent path %q normalized to %q", path, *got)
		}
	}
}

func TestShaderFixesPreservesShaderWhileAnotherModOwnsIt(t *testing.T) {
	t.Parallel()
	h := newShaderFixesHarness(t)
	first := filepath.Join(h.modsPath, "Group", "First")
	second := filepath.Join(h.modsPath, "Group", "Second")
	for _, modPath := range []string{first, second} {
		writeFile(t, filepath.Join(modPath, "ShaderFixes", "shared.ini"), "shared")
		if _, err := h.service.HandleShaders(modPath, true); err != nil {
			t.Fatal(err)
		}
	}
	if _, err := h.service.HandleShaders(first, false); err != nil {
		t.Fatal(err)
	}
	if !exists(filepath.Join(h.importerPath, "ShaderFixes", "shared.ini")) {
		t.Fatal("shared shader should remain")
	}
	if _, err := h.service.HandleShaders(second, false); err != nil {
		t.Fatal(err)
	}
	if exists(filepath.Join(h.importerPath, "ShaderFixes", "shared.ini")) {
		t.Fatal("shared shader should be removed")
	}
}

func TestShaderFixesPreservesChangedShader(t *testing.T) {
	t.Parallel()
	h := newShaderFixesHarness(t)
	modPath := filepath.Join(h.modsPath, "Group", "Mod")
	target := filepath.Join(h.importerPath, "ShaderFixes", "changed.ini")
	writeFile(t, filepath.Join(modPath, "ShaderFixes", "changed.ini"), "original")
	if _, err := h.service.HandleShaders(modPath, true); err != nil {
		t.Fatal(err)
	}
	writeFile(t, target, "changed")
	if _, err := h.service.HandleShaders(modPath, false); err != nil {
		t.Fatal(err)
	}
	raw, err := os.ReadFile(target)
	if err != nil || string(raw) != "changed" {
		t.Fatalf("preserved content = %q %v", raw, err)
	}
	if exists(filepath.Join(modPath, shaderFixesModMarkerFile)) {
		t.Fatal("mod marker should be removed")
	}
}

func TestShaderFixesMigratesManifestsByScanningCurrentImporterMods(t *testing.T) {
	t.Parallel()
	h := newShaderFixesHarness(t)
	first := filepath.Join(h.modsPath, "Group", "First")
	second := filepath.Join(h.modsPath, "Group", "Second")
	target := filepath.Join(h.importerPath, "ShaderFixes", "shared.ini")
	hash := sha256Hex("shared")
	writeFile(t, target, "shared")
	for _, item := range []struct {
		modPath, modKey string
	}{{first, "first"}, {second, "second"}} {
		writeJSON(t, filepath.Join(item.modPath, shaderFixesModMarkerFile), map[string]any{
			"version": 1,
			"modKey":  item.modKey,
			"files": []map[string]any{{
				"file": "shared.ini", "targetPath": target, "targetKey": "legacy-target-key", "hash": hash,
			}},
		})
	}
	otherImporter := filepath.Join(h.root, "OtherImporter")
	otherMods := filepath.Join(otherImporter, "Mods")
	if err := os.MkdirAll(otherMods, 0o755); err != nil {
		t.Fatal(err)
	}
	h.importers = append(h.importers, shaderImporter{Key: "OtherImporter", ImporterFolder: otherImporter})
	h.games = append(h.games, shaderGame{Game: "Other", ModFolderPath: otherMods, Importer: "OtherImporter"})
	h.service.globCalls = nil

	if _, err := h.service.HandleShaders(first, false); err != nil {
		t.Fatal(err)
	}
	assertGlobCalls(t, h.service.globCalls, []shaderGlobCall{{Pattern: "**/" + shaderFixesModMarkerFile, Cwd: h.modsPath}})
	if !exists(target) {
		t.Fatal("shared shader should remain")
	}
	assertOwnerIndex(t, filepath.Join(h.importerPath, "ShaderFixes", shaderFixesModMarkerFile), map[string]any{
		"version": float64(1),
		"targets": map[string]any{"shared.ini": map[string]any{"hash": hash, "owners": []any{"second"}}},
	})

	h.service.globCalls = nil
	if _, err := h.service.HandleShaders(second, false); err != nil {
		t.Fatal(err)
	}
	if len(h.service.globCalls) != 0 {
		t.Fatalf("second disable glob calls = %#v", h.service.globCalls)
	}
	if exists(target) {
		t.Fatal("shared shader should be removed")
	}
}

func TestShaderFixesCleansOriginalImporterAfterMove(t *testing.T) {
	t.Parallel()
	h := newShaderFixesHarness(t)
	original := filepath.Join(h.modsPath, "Group", "Mod")
	originalTarget := filepath.Join(h.importerPath, "ShaderFixes", "moved.ini")
	writeFile(t, filepath.Join(original, "ShaderFixes", "moved.ini"), "moved")
	if _, err := h.service.HandleShaders(original, true); err != nil {
		t.Fatal(err)
	}
	otherImporter := filepath.Join(h.root, "OtherImporter")
	otherMods := filepath.Join(otherImporter, "Mods")
	moved := filepath.Join(otherMods, "Group", "Mod")
	if err := os.MkdirAll(filepath.Dir(moved), 0o755); err != nil {
		t.Fatal(err)
	}
	h.importers = append(h.importers, shaderImporter{Key: "OtherImporter", ImporterFolder: otherImporter})
	h.games = append(h.games, shaderGame{Game: "Other", ModFolderPath: otherMods, Importer: "OtherImporter"})
	if err := os.Rename(original, moved); err != nil {
		t.Fatal(err)
	}
	_ = os.Remove(filepath.Join(h.importerPath, "ShaderFixes", shaderFixesModMarkerFile))
	h.service.globCalls = nil

	if _, err := h.service.HandleShaders(moved, false); err != nil {
		t.Fatal(err)
	}
	assertGlobCalls(t, h.service.globCalls, []shaderGlobCall{{Pattern: "**/" + shaderFixesModMarkerFile, Cwd: h.modsPath}})
	if exists(originalTarget) {
		t.Fatal("original shader should be removed")
	}
	if exists(filepath.Join(moved, shaderFixesModMarkerFile)) {
		t.Fatal("moved marker should be removed")
	}
	assertOwnerIndex(t, filepath.Join(h.importerPath, "ShaderFixes", shaderFixesModMarkerFile), map[string]any{
		"version": float64(1), "targets": map[string]any{},
	})
	if exists(filepath.Join(otherImporter, "ShaderFixes", shaderFixesModMarkerFile)) {
		t.Fatal("new importer should not receive an owner index")
	}
}

func TestShaderFixesRebuildsCorruptedOwnerIndex(t *testing.T) {
	t.Parallel()
	h := newShaderFixesHarness(t)
	modPath := filepath.Join(h.modsPath, "Group", "Mod")
	target := filepath.Join(h.importerPath, "ShaderFixes", "legacy.ini")
	hash := sha256Hex("legacy")
	writeFile(t, target, "legacy")
	writeJSON(t, filepath.Join(modPath, shaderFixesModMarkerFile), map[string]any{
		"version": 1,
		"modKey":  "legacy",
		"files": []map[string]any{{
			"file": "legacy.ini", "targetPath": target, "targetKey": "legacy-target-key", "hash": hash,
		}},
	})
	writeFile(t, filepath.Join(h.importerPath, "ShaderFixes", shaderFixesModMarkerFile), "invalid")
	h.service.globCalls = nil

	if _, err := h.service.HandleShaders(modPath, false); err != nil {
		t.Fatal(err)
	}
	assertGlobCalls(t, h.service.globCalls, []shaderGlobCall{{Pattern: "**/" + shaderFixesModMarkerFile, Cwd: h.modsPath}})
	if exists(target) {
		t.Fatal("legacy shader should be removed")
	}
}

func TestShaderFixesRollbackRemovesOwnerState(t *testing.T) {
	t.Parallel()
	h := newShaderFixesHarness(t)
	modPath := filepath.Join(h.modsPath, "Group", "Mod")
	target := filepath.Join(h.importerPath, "ShaderFixes", "rollback.ini")
	writeFile(t, filepath.Join(modPath, "ShaderFixes", "rollback.ini"), "rollback")
	processed, err := h.service.HandleShaders(modPath, true)
	if err != nil {
		t.Fatal(err)
	}
	if err := h.service.RollbackEnabledShaders(modPath, processed); err != nil {
		t.Fatal(err)
	}
	if exists(target) {
		t.Fatal("rolled back shader should be removed")
	}
	assertOwnerIndex(t, filepath.Join(h.importerPath, "ShaderFixes", shaderFixesModMarkerFile), map[string]any{
		"version": float64(1), "targets": map[string]any{},
	})
	if exists(filepath.Join(modPath, shaderFixesModMarkerFile)) {
		t.Fatal("mod marker should be removed")
	}
}

func TestShaderFixesRollbackAfterOwnerIndexWriteFailure(t *testing.T) {
	t.Parallel()
	h := newShaderFixesHarness(t)
	modPath := filepath.Join(h.modsPath, "Group", "Mod")
	target := filepath.Join(h.importerPath, "ShaderFixes", "rollback.ini")
	writeFile(t, filepath.Join(modPath, "ShaderFixes", "rollback.ini"), "rollback")
	h.service.failOwnerIndexWriteOn = 2
	_, err := h.service.HandleShaders(modPath, true)
	if err == nil {
		t.Fatal("expected owner index write failure")
	}
	processed := processedFilesFromError(err)
	if len(processed) != 1 {
		t.Fatalf("processedFiles = %#v", processed)
	}
	h.service.failOwnerIndexWriteOn = 0
	if err := h.service.RollbackEnabledShaders(modPath, processed); err != nil {
		t.Fatal(err)
	}
	if exists(target) {
		t.Fatal("rolled back shader should be removed")
	}
	assertOwnerIndex(t, filepath.Join(h.importerPath, "ShaderFixes", shaderFixesModMarkerFile), map[string]any{
		"version": float64(1), "targets": map[string]any{},
	})
}

func TestShaderFixesDoesNotCopyReservedOwnerIndex(t *testing.T) {
	t.Parallel()
	h := newShaderFixesHarness(t)
	modPath := filepath.Join(h.modsPath, "Group", "Mod")
	writeFile(t, filepath.Join(modPath, "ShaderFixes", "actual.ini"), "actual")
	writeFile(t, filepath.Join(modPath, "ShaderFixes", shaderFixesModMarkerFile), "reserved")
	if _, err := h.service.HandleShaders(modPath, true); err != nil {
		t.Fatal(err)
	}
	raw, err := os.ReadFile(filepath.Join(modPath, shaderFixesModMarkerFile))
	if err != nil {
		t.Fatal(err)
	}
	var manifest map[string]any
	if err := json.Unmarshal(raw, &manifest); err != nil {
		t.Fatal(err)
	}
	assertOwnerIndex(t, filepath.Join(h.importerPath, "ShaderFixes", shaderFixesModMarkerFile), map[string]any{
		"version": float64(1),
		"targets": map[string]any{
			"actual.ini": map[string]any{
				"hash":   sha256Hex("actual"),
				"owners": []any{manifest["modKey"]},
			},
		},
	})
}

func TestShaderFixesDoesNotScanWhenNoManifestExists(t *testing.T) {
	t.Parallel()
	h := newShaderFixesHarness(t)
	modPath := filepath.Join(h.modsPath, "Group", "Mod")
	writeFile(t, filepath.Join(modPath, "ShaderFixes", "unused.ini"), "unused")
	h.service.globCalls = nil
	if _, err := h.service.HandleShaders(modPath, false); err != nil {
		t.Fatal(err)
	}
	if len(h.service.globCalls) != 0 {
		t.Fatalf("glob calls = %#v, want none", h.service.globCalls)
	}
}

type shaderFixesHarness struct {
	root         string
	importerPath string
	modsPath     string
	importers    []shaderImporter
	games        []shaderGame
	service      *ShaderFixes
}

func newShaderFixesHarness(t *testing.T) *shaderFixesHarness {
	t.Helper()
	root := t.TempDir()
	importerPath := filepath.Join(root, "Importer")
	modsPath := filepath.Join(importerPath, "Mods")
	if err := os.MkdirAll(modsPath, 0o755); err != nil {
		t.Fatal(err)
	}
	h := &shaderFixesHarness{
		root: root, importerPath: importerPath, modsPath: modsPath,
		importers: []shaderImporter{{Key: "Importer", ImporterFolder: importerPath}},
		games:     []shaderGame{{Game: "Test", ModFolderPath: modsPath, Importer: "Importer"}},
	}
	h.service = newShaderFixes()
	h.service.getImporters = func() []shaderImporter { return h.importers }
	h.service.getGames = func() []shaderGame { return h.games }
	return h
}

func writeFile(t *testing.T, path, content string) {
	t.Helper()
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(path, []byte(content), 0o666); err != nil {
		t.Fatal(err)
	}
}

func writeJSON(t *testing.T, path string, value any) {
	t.Helper()
	raw, err := json.MarshalIndent(value, "", "  ")
	if err != nil {
		t.Fatal(err)
	}
	writeFile(t, path, string(raw)+"\n")
}

func exists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

func sha256Hex(value string) string {
	sum := sha256.Sum256([]byte(value))
	return hex.EncodeToString(sum[:])
}

func assertOwnerIndex(t *testing.T, path string, want map[string]any) {
	t.Helper()
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	var got map[string]any
	if err := json.Unmarshal(raw, &got); err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("owner index = %#v, want %#v", got, want)
	}
}

func assertGlobCalls(t *testing.T, got, want []shaderGlobCall) {
	t.Helper()
	if len(got) != len(want) {
		t.Fatalf("glob calls = %#v, want %#v", got, want)
	}
	for i := range want {
		if got[i].Pattern != want[i].Pattern || filepath.Clean(got[i].Cwd) != filepath.Clean(want[i].Cwd) {
			t.Fatalf("glob calls = %#v, want %#v", got, want)
		}
	}
}
